---
title: "RNA-seq Analysis"
author: "Bob Policastro"
output: html_document
---

```{r message=FALSE, warning=FALSE}
library("tidyverse")
library("DESeq2")
library("ComplexHeatmap")
library("PCAtools")
library("EnhancedVolcano")
```

### Preparing Data

#### Loading Sample Info

```{r message=FALSE}
runinfo <- "run_info.tsv" %>%
  read_tsv(col_names=c("info", "SRR_1", "SRR_2")) %>%
  separate(info, into=c("GSM", "name", "organism", "seqtype"), sep="[;:] ")

knitr::kable(runinfo)
```

#### Loading Counts

```{r message=FALSE}
counts <- "results/counts/counts.tsv" %>%
  read_tsv(skip=1) %>%
  select(!c(Chr, Start, End, Strand, Length)) %>%
  rename_with(~str_extract(.x, "(?<=aligned/)[[:alnum:]]+(?=_Aligned)"), !Geneid)

knitr::kable(counts[1:5, 1:5])
```

#### Prettier Sample Names

```{r message=FALSE}
runinfo <- runinfo %>%
  pivot_longer(starts_with("SRR"), names_to="SRR", values_to="accession") %>%
  transmute(accession, name=str_c(accession, name, sep="_"))

counts <- counts %>%
  pivot_longer(!Geneid, names_to="accession", values_to="count") %>%
  right_join(runinfo, by="accession") %>%
  select(!accession) %>%
  pivot_wider(names_from=name, values_from=count)

knitr::kable(counts[1:5, 1:5])
```

### Differential Gene Expression

#### Create Count Matrix

```{r message=FALSE}
count_matrix <- counts %>%
  column_to_rownames("Geneid") %>%
  as.matrix

knitr::kable(count_matrix[1:5, 1:5])
```

#### Create Design Table

```{r message=FALSE}
coldata <- count_matrix %>%
  colnames %>%
  {tibble(sample=.)} %>%
  separate(sample, into=c("accession", "run"), sep="(?<=^SRR[[:digit:]]{7})_") %>%
  mutate(name=str_extract(run, "[[:alnum:]_]+(?=_r[[:digit:]]$)")) %>%
  mutate(name=factor(name, levels=c("orgR_GO_f", "orgR_GO_m"))) %>%
  group_by(run) %>%
  mutate(run=str_c(name, sprintf("run_%s", cur_group_id()), sep="_")) %>%
  column_to_rownames("accession")

knitr::kable(coldata)
```

#### Differential Expression

```{r message=FALSE}
dds <- count_matrix %>%
  DESeqDataSetFromMatrix(coldata, design=~name) %>%
  collapseReplicates(coldata$run) %>%
  DESeq

degs <- dds %>%
  results %>%
  as.data.frame %>%
  rownames_to_column("gene") %>%
  as_tibble %>%
  arrange(padj)

knitr::kable(head(degs))
```

#### Normalized Counts

```{r message=FALSE}
rlog_counts <- dds %>%
  rlog(blind=FALSE) %>%
  assay %>%
  as.data.frame %>%
  rownames_to_column("gene") %>%
  as_tibble

knitr::kable(rlog_counts[1:5, 1:5])
```

### Plots

#### PCA

```{r message=FALSE}

metadata <- rlog_counts %>%
  column_to_rownames("gene") %>%
  colnames %>%
  {data.frame(type=ifelse(str_detect(., "_m_"), "male", "female"))} %>%
  magrittr::set_rownames(colnames(column_to_rownames(rlog_counts, "gene")))
  
pca_results <- rlog_counts %>%
  column_to_rownames("gene") %>%
  pca(metadata=metadata)
```

PCA biplot.

```{r message=FALSE}
biplot(pca_results, colby="type")
```

PCA loadings.

```{r message=FALSE} 
plotloadings(pca_results, rangeRetain=0.01)
```

#### Volcano

```{r message=FALSE}
EnhancedVolcano(degs, degs$gene, "log2FoldChange", "padj")
```
